New README (#105)
PolymerLabs/playground-elements: README.md

- Embed interactive code examples in your documentation.
- Build interactive tutorials and example galleries.
- Build full-featured coding sandboxes (think Glitch or JSBin).

#### 🤯 No backend required

Unlike other coding environments, Playground never sends code to a backend
server for compilation or serving. Instead, Playground uses a Service Worker to
create a virtual URL-space that runs 100% within the browser. If you can host
static files, you can host a Playground.

#### <img src="images/typescript.png" width="18" height="18"> TypeScript support

Playground automatically compiles `.ts` files using TypeScript. Compilation
happens in a Web Worker on a separate thread, so your page stays responsive.

#### 🧩 Web Components

Playground uses Web Components, so it doesn't require a framework. Yet it will
play nicely with any framework you're already using, like React, Vue, and
Angular.

#### 🍱 Mix-and-match for custom layouts

Playground is broken up into small components like an editor, file picker, and
preview. Mix-and-match components to create any layout you want, or just use
`<playground-ide>` for an easy out-of-the-box experience.

> #### 🎨 Themable

Reviewer: Do you think it's worth mentioning how you can define samples inline for quick authoring, or in files?

Also, do you want to open up the pre-renderer and mention that? Progressive enhancement is a big feature!

Playground is fully themeable with CSS Custom Properties, down to the color of
each kind of syntax-highlighted token. It comes with lots of preset themes, too.

---

<h3 align="center">
  <a href="https://polymerlabs.github.io/playground-elements/">
    Try the demo and configurator!
  </a>
</h3>

---

## Getting Started

Install Playground from NPM:

```sh
npm install --save playground-elements
```

Import the `playground-elements` module to register the Playground custom
elements:

```html
<script type="module">

>   import 'playground-elements';

Author: > Is this better than showing using `src="..."` with a reference into node_modules? I feel like most projects aren't setup to run module resolution on inline scripts, so either way is only going to work in certain situations.

Switched to `<script src="/node_modules/playground-elements/playground-ide.js"></script>`.

> Maybe show loading from Skypack?

Loading from Skypack doesn't work yet. I've got a proposed fix though, will send after this PR.

>   import 'playground-elements';

Author: Filed https://github.com/PolymerLabs/playground-elements/issues/107 explaining the problem with Skypack. Solution is pretty simple.

</script>
```

Create a `<playground-ide>` element in your HTML, and specify your project files
inline:

```html
<playground-ide editable-filesystem line-numbers resizable>
  <script type="sample/html" filename="index.html">
    <!doctype html>
    <body>
      Hello
      <script type="module" src="./index.js">&lt;/script>
    </body>
  </script>

  <script type="sample/ts" filename="index.ts">
    document.body.appendChild(document.createTextNode("World!"))
  </script>
</playground-ide>
```

Serve with any HTTP server that supports bare module import resolution, such
as [@web/dev-server](https://modern-web.dev/docs/dev-server/overview/):

```sh
npm install --save-dev @web/dev-server
npx web-dev-server --node-resolve --watch
```

You can also use a tool like [Rollup](https://rollupjs.org/guide/en/) to resolve
module imports to paths at build time. If you need more help with building and
serving, check out the [Modern Web Guides](https://modern-web.dev/guides/).

### Compatibility

Playground is supported by all modern browsers. It requires support for custom

> elements, modules, service workers, and web workers.

Author: Done.

&nbsp;&nbsp;&nbsp; **Supported:** &nbsp;&nbsp;<img src="images/check-green.png" width="20px" height="20px" class="check" alt="Supported"> <img src="images/chrome.png" width="20px" height="20px"> Chrome
&nbsp;&nbsp;&nbsp;
<img src="images/check-green.png" width="20px" height="20px" class="check" alt="Supported"> <img src="images/firefox.png" width="20px" height="20px"> Firefox
&nbsp;&nbsp;&nbsp;
<img src="images/check-green.png" width="20px" height="20px" class="check" alt="Supported"> <img src="images/safari.png" width="20px" height="20px"> Safari
&nbsp;&nbsp;&nbsp;
<img src="images/check-green.png" width="20px" height="20px" class="check" alt="Supported"> <img src="images/edge.png" width="20px" height="20px"> Edge

**Unsupported:** &nbsp; <img src="images/red-cross.png" width="20px" height="20px" class="check" alt="Unsupported"> <img src="images/ie.png" width="20px" height="20px"> Internet Explorer

## Project files

There are 3 ways to specify the files of a playground project.

### Option 1: Inline scripts

> Add one or more `<script type="sample/..." filename="...">` tags as children of

Reviewer: Can you document the supported types? Just a short list would do.

Author: Done.

Leading whitespace that is common to all lines is automatically removed unless
the `preserve-whitespace` attribute is present.

```html
<playground-project>
  <script type="sample/html" filename="index.html">
    <!doctype html>
    <head>
      <script type="module" src="javascript.js">&lt;/script>
      <script type="module" src="typescript.js">&lt;/script>
      <link rel="stylesheet" href="styles.css">
    </head>
    <body>
      <p>Hello World!</p>
    </body>
  </script>

  <script type="sample/js" filename="javascript.js">
    console.log('hello from javascript');
  </script>

  <script type="sample/ts" filename="typescript.ts">
    console.log('hello from typescript');
  </script>

  <script type="sample/css" filename="styles.css">
    body { color: blue; }
  </script>
</playground-project>
```

### Option 2: JSON manifest

Serve a JSON file containing a `files` object, with relative filenames.

```html
<playground-ide project-src="/path/to/my/project.json"> </playground-ide>
```

```json
{
  "files": {
    "index.html": {},
    "javascript.js": {},
    "typescript.ts": {},
    "styles.css": {}
  }
}
```

> ## Option 3: Files property

Author: Fixed.

> Directly set the `files` property to an array of files.

Author: Done.

```js
const ide = document.querySelector('playground-ide');
ide.files = [
  {
    name: "index.html",
    content: "<p>Hello World!</p>",
  },
  ...
],
```

If both `project-src` and `files` are set, then the one set most recently has
precedence. When either are set, inline scripts are ignored.

> ## Module resolution

Reviewer: I think it might be good to have a section on TypeScript, mainly to document what the compiler settings are and say that .ts files are imported with .js extensions like how `tsc` works.

Author: Done.

to do this to pin a specific version, change CDNs, or point to a locally served
copy of a module:

```js
{
  "files": { ... },
  "importMap": {
    "imports": {
      "lit-html": "https://cdn.skypack.dev/lit-html@^1.3.0",
      "lit-html/": "https://cdn.skypack.dev/lit-html@^1.3.0/"
    }
  }
}
```

When using inline project files, you can specify your import map like so:

```html
<playground-ide>
  <script type="sample/importmap">
    {
      "imports": {
        "lit-html": "https://cdn.skypack.dev/lit-html@^1.3.0",
        "lit-html/": "https://cdn.skypack.dev/lit-html@^1.3.0/"
      }
    }
  </script>
  ...
</playground-ide>
```

If an import map is defined, but does not contain an entry for a bare module,
then playground defaults to the `unpkg.com` URL.

## Custom layouts

`<playground-ide>` provides a complete out-of-the-box experience that's a good
start for many use-cases, but you can mix-and-match the various Playground
sub-components to make your custom layouts.

For example, say we need a layout with an editor above, a preview below, and
only one particular file from the project visible — like this:

<img src="./images/example.png">

> Import the components you need. The main `playground-elements` import loads all

Author: Done.

Playground elements, but when making a custom layout it's a good idea to only
load the sub-components you're actually using. This will make your JavaScript

> bundle smaller.

Author: Yeah, if you don't need the tab bar (which I think is a common configuration), then this lets you skip `mwc-tab-bar`, `mwc-tab`, `mwc-menu-surface`, `mwc-list`, `mwc-list-item`, `mwc-textfield`, and `mwc-button`. And our own code, of course.

| `--playground-preview-toolbar-foreground-color`    | ![](images/colors/444444.png) `#444444`                                                  | Text `color` of the preview toolbar                                                                                                |
| `--playground-border`                              | ![](images/colors/DDDDDD.png)`1px solid #DDDDDD`                                         | Outer and inner border                                                                                                             |
| `--playground-floating-controls-highlight-color`   | ![](images/colors/6200EE.png) `var(--playground-highlight-color, #6200EE)`               | Highlight color of popup controls buttons and inputs                                                                               |

## Shadow Parts

The following [CSS shadow
parts](https://developer.mozilla.org/en-US/docs/Web/CSS/::part) are exported,

> into which you can insert additional style rules not covered by the above CSS

Reviewer: ```suggestion
which you can style with additional rules not covered by the above CSS
```

Author: Done.
